nucypher/blockchain/eth/actors.py

```diff
@@ -1643,84 +1653,201 @@ def __init__(self,
+        max_bonus_tokens = self.economics.maximum_allowed_locked - self.economics.minimum_allowed_locked
+        if (min_whale_bonus_bid * bonus_worklock_supply) // bonus_eth_supply <= max_bonus_tokens:
+            raise self.BidderError(f"Internal error: At least one of bidders {whales} has allowable bid")
+
+        a = min_whale_bonus_bid * bonus_worklock_supply - max_bonus_tokens * bonus_eth_supply
+        b = bonus_worklock_supply - max_bonus_tokens * len(whales)
+        refund = -(-a//b)  # div ceil
+        min_whale_bonus_bid -= refund
+        whales = dict.fromkeys(whales.keys(), min_whale_bonus_bid)
+        self._all_bonus_bidders.update(whales)
+
+        return whales
+
+    def force_refund(self) -> dict:
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        whales = self.get_whales()
+        if not whales:
+            raise self.BidderError(f"All bids are allowable, force refund is not needed")
+
+        new_whales = whales
+        while new_whales:
+            whales.update(new_whales)
+            whales = self._reduce_bids(whales)
+            new_whales = self.get_whales()
+
+        receipt = self.worklock_agent.force_refund(checksum_address=self.checksum_address,
+                                                   addresses=list(whales.keys()))
+
+        if self.get_whales(force_read=True):
+            raise self.BidderError(f"Internal error: offline simulation differs from transaction results")
+        return receipt
+
+    # TODO better control: max iterations, interactive mode
+    def verify_bidding_correctness(self, gas_limit: int) -> dict:
+        self._ensure_cancellation_window(ensure_closed=True)
+
+        if self.worklock_agent.bidders_checked():
+            raise self.BidderError(f"Check was already done")
+
+        whales = self.get_whales()
+        if whales:
+            raise self.BidderError(f"Some bidders have bids that are too high: {whales}")
+
+        receipts = dict()
+        iteration = 1
+        while not self.worklock_agent.bidders_checked():
+            receipt = self.worklock_agent.verify_bidding_correctness(checksum_address=self.checksum_address,
+                                                                     gas_limit=gas_limit)
+            receipts[iteration] = receipt
+            iteration += 1
+        return receipts
+
     def refund_deposit(self) -> dict:
         """Refund ethers for completed work"""
         if not self.available_refund:
             raise self.BidderError(f'There is no refund available for {self.checksum_address}')
         receipt = self.worklock_agent.refund(checksum_address=self.checksum_address)
         return receipt
 
+    def withdraw_compensation(self) -> dict:
+        """Withdraw compensation after force refund"""
+        if not self.available_compensation:
+            raise self.BidderError(f'There is no compensation available for {self.checksum_address}')
```

Comment:
Perhaps more details here as an escape hatch: "...did you mean to call 'refund deposit'?"

Reply:
✔️